### DISCUSSION TITLE
Raspberry pi pico w UDP socket recv buffer issue

### DISCUSSION THREAD

**Discussion by Steve-2040** [micropython/micropython#13214] [RP2040 / Pico]
I have created a UDP socket client for the pi pico w. The UDP client is using MicroPython. The sock.recvfrom function behave unpredictable on the pi pico w. It looks like the receive UDP buffer is blocked till the sock.sendto is called. The received datagrams is lagging behind the expected arrival patten. This may be due to the WLAN only working on one core limitation but I also have tried moving the sock.recvfrom to the main thread with no lack. The UDP server reply to every message send.
MicroPython v1.21.0 on pi pico w and ESP32 and Python 3.11.4 on Windows 10
**pi pico w UDP client (don't work as expected):**
```py
import network
import socket
import time
import _thread
from machine import Timer
from machine import Pin

led = Pin("LED", Pin.OUT)

wlan = network.WLAN(network.STA_IF)
def wlan_connect():
    ssid = 'myWiFi'
    password = 'myPassword'

    if wlan.isconnected():
        print("Wi-Fi already connected")
        return

    wlan.active(True)
    wlan.connect(ssid, password)

    max_wait = 10
    while max_wait > 0:
        if wlan.status() < 0 or wlan.status() >= 3:
            break
        max_wait -= 1
        print('waiting for connection...')
        time.sleep(1)

    if wlan.status() != 3:
        raise RuntimeError('network connection failed')
    else:
        print('connected')
        status = wlan.ifconfig()
        print( 'ip = ' + status[0] )

wlan_connect()

server_addr = ('222.222.178.98', 8266)
sock = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
sock.bind(('', 0)) # only required when recvfrom called before sendto

exit_thread = False
def recv_thread():
    print(f"recv start")
    while not exit_thread:
        data, address = sock.recvfrom(1024)
        print (f"Received from server {address}: {data.decode()}")

keepalive_id = 0
def timer_20s(timer):
    global keepalive_id
    sock.sendto(("keepalive" + str(keepalive_id)).encode(), server_addr)
    keepalive_id += 1
    led.toggle()

_thread.start_new_thread(recv_thread, ())
t0 = Timer()
t0.init(period=20000, mode=Timer.PERIODIC, callback=timer_20s)

while True:
    msg = input()
    sock.sendto(msg.encode(), server_addr)
    if msg == "exit":
        break

t0.deinit()
exit_thread = True
sock.close()
wlan.deinit()
```
output server / client:
Received from client ('111.111.40.9', 56254): keepalive1     no client output 
Received from client ('111.111.40.9', 56254): keepalive2     Received from server ('222.222.178.98', 8266): keepalive1
Received from client ('111.111.40.9', 56254): keepalive3     Received from server ('222.222.178.98', 8266): keepalive2

**ESP32 UDP client (work as expected):**
```py
import network
import socket
import time
import _thread
from machine import Timer

def wlan_connect():
    ssid = 'myWiFi'
    password = 'myPassword'

    wlan = network.WLAN(network.STA_IF)
    if wlan.isconnected():
        print("Wi-Fi already connected")
        return
    
    wlan.active(True)
    wlan.connect(ssid, password)
    
    while True:
        if wlan.isconnected():
            break
        print('waiting for connection...')
        time.sleep(1)
        
    status = wlan.ifconfig()
    print( 'ip = ' + status[0] )

wlan_connect()

server_addr = ('222.222.178.98', 8266)
sock = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
sock.bind(('', 0)) # only required when recvfrom called before sendto

exit_thread = False
def recv_thread():
    print("recv start")
    while not exit_thread:
        data, address = sock.recvfrom(1024)
        print (f"Received from {address}: {data.decode()}")

keepalive_id = 0
def timer_20s(timer):
    global keepalive_id
    sock.sendto(("keepalive" + str(keepalive_id)).encode(), server_addr)
    keepalive_id += 1

_thread.start_new_thread(recv_thread, ())
t0 = Timer(0)
t0.init(period=20000, mode=Timer.PERIODIC, callback=timer_20s)

while True:
    msg = input()
    sock.sendto(msg.encode(), server_addr)
    if msg == "exit":
        break

t0.deinit()
exit_thread = True
sock.close()
#disable Wi-Fi
network.WLAN(0).active(0) #STA
network.WLAN(1).active(0) #AP
```
output server / client:
Received from client ('111.111.40.9', 50045): keepalive0    Received from server ('222.222.178.98', 8266): keepalive0
Received from client ('111.111.40.9', 50045): keepalive1    Received from server ('222.222.178.98', 8266): keepalive1
Received from client ('111.111.40.9', 50045): keepalive2    Received from server ('222.222.178.98', 8266): keepalive2
**Windows 10 UDP client (work as expected):**
```py
import socket
import time
import threading

addr = ('222.222.178.98', 8266)
sock = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
sock.bind(('', 0)) # only required when recvfrom called before sendto

def recv_thread():
    print(f"recv start")
    while True:
        data, address = sock.recvfrom(1024)
        print (f"Received from {address}: {data.decode()}")

def timer_20s():
    keepalive_id = 0
    while True:
        time.sleep(20)
        sock.sendto(("keepalive" + str(keepalive_id)).encode(), addr)
        keepalive_id += 1

threading.Thread(target = recv_thread).start()
threading.Thread(target = timer_20s).start()

while True:
    msg = input()
    sock.sendto(msg.encode(), addr)
```
output server / client:
Received from client ('111.111.40.9', 49921): keepalive0    Received from server ('222.222.178.98', 8266): keepalive0
Received from client ('111.111.40.9', 49921): keepalive1    Received from server ('222.222.178.98', 8266): keepalive1
Received from client ('111.111.40.9', 49921): keepalive2    Received from server ('222.222.178.98', 8266): keepalive2

**Comment by mendenm** [comment]
There have been quite  few reports of network issues running on the second processor.  I try to only use the thread for compute-intensive, tasks, and keep all I/O on the main processor.  This usually results in a lot better outcomes.  You might try recasting the I/O part to run on the main processor, using some asyncio bits. Note that the UDP/asyncio official code has been worked out yet, but you can just use some polls of the socket, along with await asyncio.sleep() calls, to get mostly asyncio code.

**Comment by Steve-2040** [comment]
@mendenm thanks for letting me know about the asyncio option. I will look into that option as well.
